### PR TITLE
[FIRRTL] Use set-based logic to test for layer compatibility

### PR DESCRIPTION
**Pull request by rwy7**
This PR updates the logic which checks for layer compatibility when casting, resolving, or defining refs.

## Layer Requirements for Hardware Objects

We consider the layers which must be enabled for the hardware object to exist. This is the set of layers which are either enabled by the parent module, or required by any parent layerblocks. We call these the ambient layers of a definition. 

For example:
```firrtl
module Foo enablelayer A:
  layerblock B:
    layerblock C:
      wire w : UInt<1>
```

At the definition site of `w`, the ambient layers are  `A`, `B`, and `B.C`. We also call these the required layers of `w`, because `w` will not exist unless all these layers are enabled.

## Layer Requirements for Probe Wires

Probe wires, or wires containing probes, have not only ambient layer requirements, but also explicit layer requirements, which are written in the wire's type.

For example:
```firrtl
module Foo enablelayer A:
  layerblock B:
    wire p : Probe<UInt<1>, C>
```

The wire `p` has ambient layers `A` and `B`, as well as an explicit layer requirement on `C`.  In order for this probe to be resolved, all three layers must be active. Together, the ambient layers and explicit layers requirements of a probe are called the required layers of a probe.

 ## Layer requirements for probe expressions
 
When probing a hardware object, the resulting probe requires any ambient layers at the site of the `probe`, or `firrtl.ref.send`, expression. Dominance rules tells us this is always an improper superset (⊇) of the required layers of the probed hardware.
 
For example:
```firrtl
module Foo enablelayer A:
  wire w : UInt<1>
  layerblock B:
    layerblock C:
      node n = probe(w)
```

The probed wire `w` requires layer `A`, but the  result of  `probe(w)` requires the layers `A`, `B`, and `B.C` to be enabled, in order to be resolved.

## Layer requirements for probe definitions

When we `define` a probe, we must ensure that the effective layer requirements of the source probe are met by the destination. Put another way, we cannot throw away layer requirements.

For example, If a probe escapes a layerblock via a define, then the "ambient requirements" of that source probe must be represented in the "explicit requirements" of the destination:
```firrtl
module Foo enablelayer A:
  output o : Probe<UInt<1>, B.C>
  layerblock B:
    layerblock C:
      wire w : UInt<1>
      define o = probe(w)
```

The requirements of `probe(w)` are respected, because `o` has an ambient requirement on layer `A`, and an explicit requirement on layer `B.C`, which in turn implies a requirement on layer `B` due to layer nesting.

A `define` op is only active when its ambient layers are enabled, so we must also check that the `define` op will be active any time the destination's layer requirements are met. Effectively, the define op cannot have more layer requirements than its destination. This check prevents us from writing code such as:

```firrtl
module Foo:
  output o : Probe<UInt<1>>
  node n = UInt<1>(0)
  wire w = probe(n)
  layerblock A:
    define o = w // bad
```

## Layer requirements for reading a probe

To read a probe, the ambient layers of the read expression, or `firrtl.ref.resolve`, must meet the effective layer requirements of the input probe. 

For example, the following is valid:
```firrtl
extmodule Foo enablelayer B:
  output p : Probe<UInt<1>, A>

module Bar enablelayer A:
  inst foo of Foo
  layerblock B:
    node n = resolve(foo.p)
```
  
## Probe color conversion rules

In the FIRRTL language, probes are implicitly converted.  This is achieved in the FIRRTL IR through an explicit cast operation.

When casting a ref, the effective layer requirements of the output must at least meet the effective layer requirements of the input probe.  The cast is allowed to do the following conversions:

1. It may drop any explicit layer requirement on the input probe that is present in the ambient layers of the cast operation.  This is because the effective layer requirements remain unchanged. For example:
  ```firrtl
  module Foo enablelayer A:
    input in : Probe<UInt<1>, A>
    output out : Probe<UInt<1>>
    define out = in
  ```

2. It may add any explicit layer requirement, with no constraint.  This is because when a probe is valid, it will still be valid under additional layer requirements. For example:
  ```firrtl
  module Foo:
    input in : Probe<UInt<1>>
    output out : Probe<UInt<1>, A>
    define out = in
  ```
 
---

Note: although it is certainly possible for a probe to effectively require multiple unrelated layers, it's not actually possible to write this as an explicit probe color. E.g., the following is not possible:
```
extmodule Foo:
  output p : Probe<UInt<1>, A>
 
module Foo:
  output p : Probe<UInt<1>, {A, B}> ;; impossible.
  layerblock B:
     inst foo of Foo
     define p = foo.p
 ```
 
 ---
 
 Note, this PR attempts to verify the validity of probes without modifying the IR or requiring that the layer requirements of probes are canonicalized / expressed explicitly in the IR, which would actually break the fir emitter. As it is, an input firrtl program should round trip "just fine" through the emitter.

### CODE REVIEW

File: test/Dialect/FIRRTL/layers-basic.mlir [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+  }
+
+  firrtl.layer @B bind {}
+
+  firrtl.module @Foo(out %o : !firrtl.probe<uint<1>, @A>) {
+    %c = firrtl.constant 0 : !firrtl.uint<1>
+    %r = firrtl.ref.send %c : !firrtl.uint<1>
+    %s = firrtl.ref.cast %r : (!firrtl.probe<uint<1>>) -> !firrtl.probe<uint<1>, @A>
+    firrtl.ref.define %o, %s : !firrtl.probe<uint<1>, @A>    
+  }
```
Comment (seldridge) [inline test/Dialect/FIRRTL/layers-basic.mlir:21]: This seems wrong? Why is this able to cast  or define outside of a `layerblock @A`, i.e., this seems like it needs an ambient layer of `A` which it doesn't have?

The rule may be that the cast can go to any layer that is legal based on the ambient layers of the cast.

I'm also surprised that the define is allowed here for the same reasons.

File: test/Dialect/FIRRTL/layers-basic.mlir [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+  firrtl.module @UncoloredDefineUnderLayerblock(out %o : !firrtl.probe<uint<1>>) {
+    // Safe to define an uncolored ref under a layerblock, as long as the define
+    // would be valid outside the layerblock..
+    %c = firrtl.constant 0 : !firrtl.uint<1>
+    firrtl.layerblock @A {
+      %r = firrtl.ref.send %c : !firrtl.uint<1>
+      firrtl.ref.define %o, %r : !firrtl.probe<uint<1>>
+    }
+  }
```
Comment (seldridge) [inline test/Dialect/FIRRTL/layers-basic.mlir:67]: I don't think this should be allowed given how defines work (there can be only one define to a probe in the entire module). The stricter check here should be that the ambient layer of a define must exactly match the explicit layer of the destination probe type.

I realize that this conflicts with the code that already landed. However, I've tried to really restrict this in the FIRRTL spec PR.

File: test/Dialect/FIRRTL/layers-basic.mlir [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+  firrtl.module @DefineUncoloredProbeUnderEnabledLayerWithReferentUnderLayerBlock(out %o : !firrtl.probe<uint<1>>)
+  attributes {layers = [@A]} {
+    firrtl.layerblock @A {
+      %c = firrtl.constant 0 : !firrtl.uint<1>
+      %r = firrtl.ref.send %c : !firrtl.uint<1>
+      firrtl.ref.define %o, %r : !firrtl.probe<uint<1>>
+    }
+  }
```
Comment (seldridge) [inline test/Dialect/FIRRTL/layers-basic.mlir:135]: I'm glad this works and should be allowed. It would be good to, in a follow-up, add a pass that can inline this layerblock.

**Review by seldridge — COMMENTED** [review]
This is coming together well! Minor nits.

The tests mostly make sense. Several did not. Specifically, it seems like this is allowing a cast or a define from an ambient uncolored layer to a colored probe. This seems wrong unless it is under a layerblock.

File: test/Dialect/FIRRTL/layers-basic.mlir (hunk 1 above)
Comment (rwy7) [reply test/Dialect/FIRRTL/layers-basic.mlir:21]: The idea is that it's okay to add a layer requirement to a probe. If a probe is valid without layer A, then it should still be valid if layer A is enabled.

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 4]
```diff
@@ -320,6 +321,170 @@ void getAsmBlockArgumentNamesImpl(Operation *op, mlir::Region &region,
...
+    if (auto module = dyn_cast<FModuleLike>(op)) {
+      for (auto layer : module.getLayers())
+        result.insert(cast<SymbolRefAttr>(layer));
+      continue;
+    }
+    if (auto layerblock = dyn_cast<LayerBlockOp>(op)) {
+      result.insert(layerblock.getLayerName());
+      continue;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (rwy7) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp:377]: > This is safe so long as all the ops are verified to be in a layerblock or a module (which is already checked!)

Just looked around in circt, I don't see this being verified for any expression op (including ref.sub, ref.send), nor most statement op (excepting layerblock, but including ref.define and when). We _are_ verifying the parent of declaration ops, though.

So, this code passes the verifiers today:
```mlir
builtin.module {
  %c = firrtl.constant 0  : !firrtl.uint<1>
  %p = firrtl.ref.send %c : !firrtl.uint<1>
}
```

I think it's reasonable to be a bit defensive here and leave the nullptr check in, but I will break when we hit the first module-like parent. Sound okay?

[the remark this answers is not in the document]
Comment (seldridge) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp:377]: Hmm... I'm surprised these checks aren't in there. There are some spotty checks that different FIRRTL ops have an `FModuleOp` parent, however, this is by no means consistent. I'd consider tightening this up so that all ops can only exist under `FModuleOp`, `WhenOp`, or `LayerBlockOp`. This may break other users or passes that are relying on this permissiveness.

Yes, this is fine as-is, then.

@nandor: are you aware of users needing to put FIRRTL operations in non-FIRRTL containers?

File: test/Dialect/FIRRTL/layers-basic.mlir (hunk 1 above)
Comment (seldridge) [reply test/Dialect/FIRRTL/layers-basic.mlir:21]: Yep! This is equivalent to saying that if a probe is conditionally defined, it is also fine to make it unconditionally defined. Equivalently, a `layerblock` can always be inlined into its parent.

File: test/Dialect/FIRRTL/layers-basic.mlir (hunk 2 above)
Comment (rwy7) [reply test/Dialect/FIRRTL/layers-basic.mlir]: We now check that the ref.define op doesn't have _more_ layer requirements than the destination.

**Review by youngar — APPROVED** [review]
LGTM

**Comment by youngar** [comment]
I think we should throw the explanation from the PR text in to a rationale document, or maybe it makes a good start for the FIRRTL spec.

**Comment by seldridge** [comment]
@youngar wrote:
> I think we should throw the explanation from the PR text in to a rationale document, or maybe it makes a good start for the FIRRTL spec.

Agree. The language and concepts introduced here "explicit", "ambient", etc. are better than what exist in the FIRRTL spec PR to explain all this (https://github.com/chipsalliance/firrtl-spec/pull/160).

File: test/Dialect/FIRRTL/layers-errors.mlir [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  firrtl.module @Top(out %o : !firrtl.probe<uint<1>>) {
+    firrtl.layerblock @A {
+      %c = firrtl.constant 0  : !firrtl.uint<1>
+      %r = firrtl.ref.send %c : !firrtl.uint<1>
+      // expected-error @below {{'firrtl.ref.define' op has more layer requirements than destination}}
+      // expected-note  @below {{additional layers required: @A}}
+      firrtl.ref.define %o, %r : !firrtl.probe<uint<1>>
+    }
+  }
+}
```
Comment (seldridge) [inline test/Dialect/FIRRTL/layers-errors.mlir:41]: Is this test checking anything given that the verification is only happening at the `firrtl.ref.define` and the `firrtl.ref.send` doesn't factor into the check? I may be missing something.

File: test/Dialect/FIRRTL/layers-errors.mlir [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+firrtl.circuit "Top" {
+  firrtl.layer @A bind {}
+  firrtl.layer @B bind {}
+  firrtl.module @Foo(out %o : !firrtl.probe<uint<1>, @B>) {
+    firrtl.layerblock @B {
+      %c = firrtl.constant 0 : !firrtl.uint<1>
+      %r = firrtl.ref.send %c : !firrtl.uint<1>
+      %s = firrtl.ref.cast %r : (!firrtl.probe<uint<1>>) -> !firrtl.probe<uint<1>, @B>
+      firrtl.ref.define %o, %s : !firrtl.probe<uint<1>, @B>
+    }
```
Comment (seldridge) [inline test/Dialect/FIRRTL/layers-errors.mlir:54]: Nit: can this be removed for test compactness and maintenance?

Further nit: the actual error logic may not differentiate from an instance vs. a wire of the same type. Consider simplifying this. I'm of two opinions of this as this is useful to check, however, it may not actually add coverage to the test. Same thing with the `ref.subop` test.

File: test/Dialect/FIRRTL/layers-errors.mlir [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+// -----
+
+firrtl.circuit "Top" {
+  firrtl.layer @A bind {}
+  firrtl.extmodule @Foo(out o : !firrtl.probe<uint<1>>) attributes {layer = [@A]}
+  firrtl.module @Top(out %o : !firrtl.probe<uint<1>>) {
+    %foo_o = firrtl.instance foo @Foo(out o : !firrtl.probe<uint<1>>)
+    firrtl.ref.define %o, %foo_o : !firrtl.probe<uint<1>>
+  }
+}
```
Comment (seldridge) [inline test/Dialect/FIRRTL/layers-errors.mlir:75]: Nit: Would this be better placed in `layers.mlir` instead of `layers-errors.mlir`?

File: test/Dialect/FIRRTL/layers-errors.mlir [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    firrtl.layerblock @A {
+      %f = firrtl.opensubfield %w[f] : !firrtl.openbundle<f: probe<uint<1>>>
+      %c = firrtl.constant 0 : !firrtl.uint<1>
+      %r = firrtl.ref.send %c : !firrtl.uint<1>
+      // expected-error @below {{'firrtl.ref.define' op has more layer requirements than destination}}
+      // expected-note  @below {{additional layers required: @A}}
+      firrtl.ref.define %f, %r : !firrtl.probe<uint<1>>
+    }
+  }
+}
```
Comment (seldridge) [inline test/Dialect/FIRRTL/layers-errors.mlir:92]: This could use at least one `ref.resolve` test.

**Review by seldridge — APPROVED** [review]
This looks excellent. Thank you for taking the time to polish the PR and sort out exactly what the sane semantics of this are.

Some comments and nits throughout. However, all of this shouldn't block merging.

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 9]
```diff
@@ -320,6 +321,128 @@ void getAsmBlockArgumentNamesImpl(Operation *op, mlir::Region &region,
...
+        return true;
+      if (cmp == 1)
+        return false;
+    }
+    return lhsNestedSize < rhsNestedSize;
+  }
+};
+} // namespace
+
+using LayerSet = SmallSet<SymbolRefAttr, 4, CompareSymbolRefAttr>;
```
[the remark this answers is not in the document]
Comment (rwy7) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp:354]: > Specifically, are the first three redundant informationally? I would think yes.

Yes!

> This does hit the "fast lookup path" in isLayerCompatible but only when stored in such a fashion (which is not guaranteed given how enablelayer is pulled in with, e.g., Y.Z, and will hit the "slow path").

And how the explicit layer requirements of a probe type are pulled in.

> This may motivate a slightly more complex data structure that avoids redundant storage while still making lookup fast.

Yep!

There are two things I like about how I've implemented it:

1) when we tell the user what layers are missing, we only tell them the missing layers that were actually written in the source. So if a missing layer requirement comes from nested layerblocks A and B, we'll list both A and A.B, if it comes from a probe color like A.B, we only list A.B.

2) I wrote this code so that it would be easier to make changes later on, in particular I kept an eye on what it would take to remove nested layer declarations.

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 10]
```diff
@@ -3496,7 +3570,21 @@ LogicalResult RefDefineOp::verify() {
...
   }
 
-  return verifyLayer(*this, getDest().getType().getLayer());
+  // This define is only enabled when its ambient layers are active. Check
+  // that whenever the destination's layer requirements are met, that this
+  // op is enabled.
+  auto ambientLayers = getAmbientLayersAt(getOperation());
+  auto dstLayers = getLayersFor(getDest());
+  SmallVector<SymbolRefAttr> missingLayers;
+  if (!isLayerSetCompatibleWith(ambientLayers, dstLayers, missingLayers)) {
```
[the remark this answers is not in the document]
Comment (rwy7) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp:3579]: I did it this way, even though it's the last check, so it's easier to add more checking to the verifier below.

File: test/Dialect/FIRRTL/layers-errors.mlir (hunk 6 above)
Comment (rwy7) [reply test/Dialect/FIRRTL/layers-errors.mlir:54]: I've changed to an ext module.

My (weakly held) opinion: this is a bit of a behavioural test. We can't test all the things, but it feels reasonable to take wires and instances as a representative subset of our declaration ops, and check that the verifier works for them. We're not just looking for a minimum set of tests that give us coverage based on how the verifier works today.

File: test/Dialect/FIRRTL/layers-errors.mlir (hunk 7 above)
Comment (rwy7) [reply test/Dialect/FIRRTL/layers-errors.mlir:75]: Removed. this shouldn't have been committed.

File: test/Dialect/FIRRTL/layers-errors.mlir (hunk 8 above)
Comment (rwy7) [reply test/Dialect/FIRRTL/layers-errors.mlir:92]: I can add some tests here but they will just be duplicates of preexisting ref.resolve tests in the more general errors.mlir https://github.com/llvm/circt/pull/6643/files#diff-e012a404f9fde3c2af947cbfbf8abeb5c8126f24a6f45d44d530e4923edb0510R1909

I decided not to move any tests around for this PR.

File: test/Dialect/FIRRTL/layers-errors.mlir (hunk 6 above)
Comment (seldridge) [reply test/Dialect/FIRRTL/layers-errors.mlir]: This is a reasonable stance to me. Instances have to work and it's reasonable to lock them in with a test.

File: test/Dialect/FIRRTL/layers-errors.mlir (hunk 5 above)
Comment (rwy7) [reply test/Dialect/FIRRTL/layers-errors.mlir:42]: This test is really here to assert that the location of the src ref doesn't matter.

**Comment by rwy7** [comment]
Thanks for the great reviews!
